From 91657bd52f0b6e9d1b655bc43c125da3b9f36fe7 Mon Sep 17 00:00:00 2001
From: Marat Stanichenko <mstanichenko@openvz.org>
Date: Mon, 30 Jun 2008 13:48:49 +0400
Subject: [PATCH 096/103] VE do open check perm

Don't check permissions in do_open if we already have checked them.
We might come into do_open from __blkdev_get which is called from do_open.
This means that we have already passed the previous test get_device_perms_ve()
and allowed an access to device.
So we don't call get_device_perms_ve() twice because we mightn't pass the test.
---
 fs/block_dev.c |   14 ++++++++++----
 1 files changed, 10 insertions(+), 4 deletions(-)

diff --git a/fs/block_dev.c b/fs/block_dev.c
index 1ed4a29..f1947df 100644
--- a/fs/block_dev.c
+++ b/fs/block_dev.c
@@ -1128,10 +1128,16 @@ static int do_open(struct block_device *bdev, struct file *file, int for_part)
 	int ret;
 	int part;
 
-	ret = get_device_perms_ve(S_IFBLK, bdev->bd_dev,
-				  file->f_mode & (FMODE_READ | FMODE_WRITE));
-	if (ret)
-	        return ret;
+	/*
+	 * We don't check permissions if we already have checked them 
+	 * and come here from __blkdev_get which is called below
+	 */
+	if (!for_part) {
+		ret = get_device_perms_ve(S_IFBLK, bdev->bd_dev,
+				  	file->f_mode & (FMODE_READ | FMODE_WRITE));
+		if (ret)
+	        	return ret;
+	}
 
 	ret = -ENXIO;
 	file->f_mapping = bdev->bd_inode->i_mapping;
-- 
1.5.4.3

